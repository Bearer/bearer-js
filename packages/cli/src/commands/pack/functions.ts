import * as archiver from 'archiver'
import * as fs from 'fs-extra'
import * as path from 'path'

import BaseCommand from '../../base-command'
import { RequireIntegrationFolder } from '../../utils/decorators'
import prepareConfig from '../../utils/prepare-config'

const CONFIG_FILE = 'bearer.config.json'
const HANDLER_NAME = 'index'
const HANDLER_NAME_WITH_EXT = [HANDLER_NAME, 'js'].join('.')

export default class PackFunctions extends BaseCommand {
  static description = 'Pack integration functions'
  static hidden = true
  static flags = {
    ...BaseCommand.flags
  }

  static args = [{ name: 'ARCHIVE_PATH', required: true }]

  @RequireIntegrationFolder()
  async run() {
    // we are assuming prepare and build have been run previously
    const { args } = this.parse(PackFunctions)
    const { ARCHIVE_PATH } = args
    const target = path.resolve(ARCHIVE_PATH)
    this.debug(`Zipping to: ${target}`)

    const output = fs.createWriteStream(target)
    const archive = archiver('zip', {
      zlib: { level: 9 }
    })
    const archiveEntries: archiver.EntryData[] = []
    archive.pipe(output)

    const archived = new Promise<archiver.EntryData[]>((resolve, reject) => {
      output.on('close', () => {
        resolve(archiveEntries)
      })
      archive.on('error', (err: any) => {
        reject(err)
      })

      archive.on('entry', (entry: archiver.EntryData) => {
        if (!entry.stats || !entry.stats.isDirectory()) {
          archiveEntries.push(entry)
        }
      })

      archive.on('warning', (err: any) => {
        this.debug(err)
      })
    })

    try {
      // add functions
      archive.directory(this.locator.buildFunctionsDir, false)
      // add CONFIG
      const functions = await this.retrieveFunctions()

      const { config, handlers } = buildLambdaDefinitions(functions)
      const finalConfig = await this.retrieveAuthConfig(config)

      archive.append(JSON.stringify(finalConfig, null, 2), { name: CONFIG_FILE })

      // add handlers
      this.debug(`Generated config: ${JSON.stringify(config, null, 2)}`)
      archive.append(handlers, { name: HANDLER_NAME_WITH_EXT })

      // ZIP
      archive.finalize()

      const entries = await archived

      this.debug(`Zip content: ${entries.length} files\n  * ${entries.map(e => e.name).join('\n  * ')}`)
      this.success('Successfully generated lambda package')
      // log files added to zip
    } catch (e) {
      this.error(e)
    }
  }

  async retrieveAuthConfig(config: { functions: Record<string, string>[] }): Promise<TBearerConfig> {
    // generate config
    const content = await fs.readFile(this.locator.authConfigPath, { encoding: 'utf8' })
    return {
      ...config,
      auth: JSON.parse(content)
    }
  }

  // TODO: rewrite this using TS AST
  async retrieveFunctions(): Promise<TFunctionNames> {
    try {
      const config = await prepareConfig(
        this.locator.authConfigPath,
        this.bearerConfig.integrationUuid,
        this.locator.srcFunctionsDir
      )
      return config.functions
    } catch (e) {
      throw e
    }
  }
}

type TFunctionNames = string[]

type TBearerConfig = {
  functions: Record<string, string>[]
  auth?: any
}

type TLambdaDefinition = {
  config: {
    functions: Record<string, string>[]
  }
  handlers: string
}

export function buildLambdaDefinitions(functions: TFunctionNames): TLambdaDefinition {
  return {
    config: {
      functions: functions.reduce(
        (acc, functionName) => {
          acc.push({ [functionName]: [HANDLER_NAME, functionName].join('.') })
          return acc
        },
        [] as Record<string, string>[]
      )
    },
    handlers: buildLambdaIndex(functions)
  }
}

function buildLambdaIndex(functions: TFunctionNames): string {
  return functions
    .map((func, index) => {
      const funcConstName = `func${index}`
      return `const ${funcConstName} = require("./dist/${func}").default;
module.exports['${func}'] = ${funcConstName}.init();
`
    })
    .join('\n')
}
