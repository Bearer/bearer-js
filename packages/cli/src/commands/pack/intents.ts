import * as archiver from 'archiver'
import * as fs from 'fs-extra'
import * as path from 'path'

import BaseCommand from '../../base-command'
import { RequireScenarioFolder } from '../../utils/decorators'
import prepareConfig from '../../utils/prepare-config'

const CONFIG_FILE = 'bearer.config.json'
const HANDLER_NAME = 'index'
const HANDLER_NAME_WITH_EXT = [HANDLER_NAME, 'js'].join('.')

export default class PackIntents extends BaseCommand {
  static description = 'Pack scenario intents'
  static hidden = true
  static flags = {
    ...BaseCommand.flags
  }

  static args = [{ name: 'ARCHIVE_PATH', required: true }]

  @RequireScenarioFolder()
  async run() {
    // we are assuming prepare and build have been run previously
    const { args } = this.parse(PackIntents)
    const { ARCHIVE_PATH } = args
    const target = path.resolve(ARCHIVE_PATH)
    this.debug(`Zipping to: ${target}`)

    const output = fs.createWriteStream(target)
    const archive = archiver('zip', {
      zlib: { level: 9 }
    })
    const archiveEntries: Array<archiver.EntryData> = []
    archive.pipe(output)

    const archived = new Promise<Array<archiver.EntryData>>((resolve, reject) => {
      output.on('close', () => {
        resolve(archiveEntries)
      })
      archive.on('error', (err: any) => {
        reject(err)
      })

      archive.on('entry', (entry: archiver.EntryData) => {
        if (!entry.stats || !entry.stats.isDirectory()) {
          archiveEntries.push(entry)
        }
      })

      archive.on('warning', (err: any) => {
        this.debug(err)
      })
    })

    try {
      // add intents
      archive.directory(this.locator.buildIntentsDir, false)
      // add CONFIG
      const intents = await this.retrieveIntents()

      const { config, handlers } = buildLambdaDefinitions(intents)
      const finalConfig = await this.retrieveAuthConfig(config)

      archive.append(JSON.stringify(finalConfig, null, 2), { name: CONFIG_FILE })

      // add handlers
      this.debug(`Generated config: ${JSON.stringify(config, null, 2)}`)
      archive.append(handlers, { name: HANDLER_NAME_WITH_EXT })

      // ZIP
      archive.finalize()

      const entries = await archived

      this.debug(`Zip content: ${entries.length} files\n  * ${entries.map(e => e.name).join('\n  * ')}`)
      this.success('Successfully generated lambda package')
      // log files added to zip
    } catch (e) {
      this.error(e)
    }
  }

  async retrieveAuthConfig(config: { intents: Record<string, string>[] }): Promise<TBearerConfig> {
    // generate config
    const content = await fs.readFile(this.locator.authConfigPath, { encoding: 'utf8' })
    return {
      ...config,
      auth: JSON.parse(content)
    }
  }

  // TODO: rewrite this using TS AST
  async retrieveIntents(): Promise<TIntentNames> {
    try {
      const config = await prepareConfig(
        this.locator.authConfigPath,
        this.bearerConfig.scenarioUuid,
        this.locator.srcIntentsDir
      )
      return config.intents
    } catch (e) {
      throw e
    }
  }
}

type TIntentNames = string[]

type TBearerConfig = {
  intents: Record<string, string>[]
  auth?: any
}

type TLambdaDefinition = {
  config: {
    intents: Record<string, string>[]
  }
  handlers: string
}

export function buildLambdaDefinitions(intents: TIntentNames): TLambdaDefinition {
  return {
    config: {
      intents: intents.reduce(
        (acc, intentName) => {
          acc.push({ [intentName]: [HANDLER_NAME, intentName].join('.') })
          return acc
        },
        [] as Record<string, string>[]
      )
    },
    handlers: buildLambdaIndex(intents)
  }
}

function buildLambdaIndex(intents: TIntentNames): string {
  return intents
    .map((intent, index) => {
      const intentConstName = `intent${index}`
      const intentInstance = `${intentConstName}Instance`
      return `const ${intentConstName} = require("./dist/${intent}").default;
const ${intentInstance} = new ${intentConstName}()
module.exports['${intent}'] = ${intentInstance}.intentType.intent(${intentInstance}.action);
`
    })
    .join('\n')
}
