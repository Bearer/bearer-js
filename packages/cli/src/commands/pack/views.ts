import { flags } from '@oclif/command'
import { spawn } from 'child_process'

import BaseCommand from '../../base-command'
import { IntegrationBuildEnv } from '../../types'
import { RequireIntegrationFolder, skipIfNoViews, RequireLinkedIntegration } from '../../utils/decorators'

const cdnHost = 'cdn-host'

export default class PackViews extends BaseCommand {
  static description = 'Pack integration views'
  static hidden = true
  static flags = {
    ...BaseCommand.flags,
    [cdnHost]: flags.string({
      required: true,
      description: 'Host url where views are uploade to (ex: https:static.bearer.sh/123456/attach-pull/dist/78901/'
    })
  }

  @skipIfNoViews()
  @RequireIntegrationFolder()
  @RequireLinkedIntegration()
  async run() {
    const { flags } = this.parse(PackViews)

    const config = this.constants
    const env: IntegrationBuildEnv = {
      ...process.env,
      BEARER_INTEGRATION_ID: this.bearerConfig.BUID!,
      BEARER_INTEGRATION_HOST: config.IntegrationServiceHost,
      BEARER_AUTHORIZATION_HOST: config.IntegrationServiceHost,
      CDN_HOST: flags[cdnHost]
    }

    try {
      const buildDestination = await this.buildStencil(env)
      this.success(`Packed views : ${buildDestination}`)
    } catch (e) {
      this.error(e)
    }
  }

  async buildStencil(env: IntegrationBuildEnv) {
    return new Promise((resolve, reject) => {
      const build = spawn('yarn', ['stencil', 'build'], { env, cwd: this.locator.buildViewsDir })

      build.stdout.on('data', data => {
        this.debug(`build integration => stdout: ${data}`)
      })

      build.stderr.on('data', data => {
        this.debug(`build integration => stderr: ${data}`)
      })

      build.on('close', code => {
        if (code === 0) {
          resolve(this.locator.buildViewsDir)
        } else {
          reject(new Error("Can't build integration views. please check logs"))
        }
      })
    })
  }
}
