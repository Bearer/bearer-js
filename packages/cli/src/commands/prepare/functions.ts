import BaseCommand from '../../base-command'
import { RequireIntegrationFolder } from '../../utils/decorators'
import { ensureFolderExists } from '../../utils/helpers'

export default class PrepareFunctions extends BaseCommand {
  static description = 'describe the command here'
  // TODO: remove prepare:intents when people have migrated to function
  static aliases = ['prepare:intents', 'p:f']
  static hidden = true
  static flags = {
    ...BaseCommand.flags
  }

  static args = []

  @RequireIntegrationFolder()
  async run() {
    ensureFolderExists(this.locator.buildFunctionsDir, true)
    ensureFolderExists(this.locator.buildArtifactDir, true)
  }
}
