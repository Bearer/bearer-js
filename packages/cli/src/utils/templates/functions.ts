import { Authentications } from '@bearer/types/lib/authentications'
import FunctionType from '@bearer/types/lib/function-types'
import * as Case from 'case'
import * as path from 'path'

import { copyFiles } from '../helpers'
import baseCommand from '../../base-command'

export default async (command: baseCommand, auth: Authentications, type: FunctionType, name: string) => {
  const vars = getVars(name, auth, type)
  return await copyFiles(command, path.join(`generate/function`, auth, type), command.locator.srcFunctionsDir, vars)
}

function getVars(name: string, authType: Authentications, functionType: FunctionType) {
  return {
    authType,
    functionType,
    fileName: name,
    functionClassName: Case.pascal(name)
  }
}
