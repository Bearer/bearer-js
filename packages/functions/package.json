{
  "private": false,
  "name": "@bearer/functions",
  "version": "1.12.0",
  "description": "Bearer Functions",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "author": "Bearer Team<engineering@bearer.sh>",
  "license": "MIT",
  "repository": "Bearer/bearer",
  "homepage": "https://github.com/Bearer/bearer/packages/functions#readme",
  "files": [
    "lib/"
  ],
  "scripts": {
    "clean": "rm -rf lib/",
    "build": "tsc -p tsconfig.build.json",
    "start": "yarn clean && yarn build -w",
    "prepare": "yarn clean && yarn build",
    "prepack": "NODE_ENV=production yarn build",
    "test": "jest --colors",
    "test:ci": "jest --coverage --runInBand --colors"
  },
  "dependencies": {
    "@bearer/logger": "^1.7.0",
    "axios": "^0.18.0"
  },
  "devDependencies": {
    "@types/jest": "^24.0.11",
    "@types/node": "^10.12.18",
    "jest": "^24.7.1",
    "jest-mock-axios": "^2.3.0",
    "ts-jest": "^24.0.1",
    "typescript": "^3.4.1"
  },
  "gitHead": "cb2654b19000e2ad6f31a0e4ed715a6298c8df2d"
}
