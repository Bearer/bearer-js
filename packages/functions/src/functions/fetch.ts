import debug from '@bearer/logger'

import * as d from '../declaration'
import { eventAsActionParams } from './utils'

const logger = debug('functions:fetch-state')

export abstract class FetchData<ReturnedData = any, TError = any, AuthContext = any> {
  // expected implementation
  abstract async action(
    event: d.TFetchActionEvent<any, AuthContext, any>
  ): Promise<d.TFetchPayload<ReturnedData, TError>>

  // Internal
  static call(action: d.TFetchAction) {
    return async (event: d.TLambdaEvent) => {
      try {
        const { error, data }: d.TFetchPayload<any, any> = await action(eventAsActionParams(event))
        if (error) {
          logger(error)
          return { error }
        }
        return { data }
      } catch (error) {
        logger.extend('FetchActionExecutionError')(error)
        throw new FetchActionExecutionError(error)
      }
    }
  }

  static init() {
    return FetchData.call(new (this.prototype.constructor as any)().action)
  }
}

export class FetchActionExecutionError extends Error {}
