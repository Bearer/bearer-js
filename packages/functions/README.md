# @bearer/functions

## Usage

**Creating a FetchData function**

```
import { FetchData, TOAUTH2AuthContext } from '@bearer/functions'
import Client from './client'

export default class ListPullRequestsFunction {
  static functionName: string = 'listPullRequests'
  static functionType: any = FetchData

  static async action({ context, params }: { context: TOAUTH2AuthContext; params: any }) {
    try {
      const { data } = await Client(context.authAccess.accessToken).get(`/repos/${params.fullName}/pulls`, {
        params: { per_page: 10, ...params }
      })
      return { data }
    } catch (error) {
      return { error: error.toString() }
    }
  }
}

```

**Creating a SaveData function**

```
import { SaveState, TOAUTH2AuthContext } from '@bearer/functions'
import Client from './client'

export type TState = {
  pullRequests: any
}

export type TParams = {
  fullName: string,
  page?: number
}

export default class ListPullRequestsFunction {
  static functionName: string = 'listPullRequests'
  static functionType: any = SaveState

  static async action({ context, params, state }: { context: TOAUTH2AuthContext; params: TParams, state: TState }) {
    try {
      const { data } = await Client(context.authAccess.accessToken).get(`/repos/${params.fullName}/pulls`, {
        params: { per_page: 10, ...params }
      })
      return { data, state: { pullRequests: data } }
    } catch (error) {
      return { error: error.toString() }
    }
  }
}

```
