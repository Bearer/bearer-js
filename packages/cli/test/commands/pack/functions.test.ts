import { buildLambdaDefinitions } from '../../../src/commands/pack/functions'

describe('pack:functions', () => {
  describe('lambdas declaration', () => {
    const { config, handlers } = buildLambdaDefinitions(['my-function', 'spongeBobFunction'])

    it('generates a correct handlers', () => {
      expect(handlers).toMatchSnapshot()
    })

    it('build functions bearer config', () => {
      expect(config).toEqual({
        functions: [
          {
            'my-function': 'index.my-function'
          },
          {
            spongeBobFunction: 'index.spongeBobFunction'
          }
        ]
      })
    })
  })
})
