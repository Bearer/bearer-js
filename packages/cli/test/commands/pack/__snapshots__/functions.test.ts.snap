// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pack:functions lambdas declaration generates a correct handlers 1`] = `
"const captureRequest = require('@bearer/x-ray').default
captureRequest()
const func0 = require(\\"./dist/my-function\\").default;
module.exports['my-function'] = func0.init();

const func1 = require(\\"./dist/spongeBobFunction\\").default;
module.exports['spongeBobFunction'] = func1.init();

  "
`;
